Filter null file collection items in `UnresolvedItemsCollector.replace`

I upgraded [AbstractCompile#getClasspath()](https://github.com/gradle/gradle/blob/9f8dcb6e8e44766364eca91f0e27a253d3d6e970/platforms/jvm/language-jvm/src/main/java/org/gradle/api/tasks/compile/AbstractCompile.java#L60) to ConfigurableFileCollection but now this test fails for me: [DeclarativeAgpSmokeSpec.groovy#L29](https://github.com/gradle/gradle/blob/5daf80312c72d6f0de9dd837aba63523b32b88b0/platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/agp/DeclarativeAgpSmokeSpec.groovy#L29) with:
```
Caused by: java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:903)
	at com.google.common.collect.ImmutableList$Builder.add(ImmutableList.java:815)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection$UnresolvedItemsCollector.replace(DefaultConfigurableFileCollection.java:683)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.replace(DefaultConfigurableFileCollection.java:373)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection$UnresolvedItemsCollector.addItem(DefaultConfigurableFileCollection.java:656)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection$UnresolvedItemsCollector.setFrom(DefaultConfigurableFileCollection.java:630)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection$UnresolvedItemsCollector.<init>(DefaultConfigurableFileCollection.java:578)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection$EmptyCollector.setFrom(DefaultConfigurableFileCollection.java:543)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.newValue(DefaultConfigurableFileCollection.java:359)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.newExplicitValue(DefaultConfigurableFileCollection.java:330)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.setFrom(DefaultConfigurableFileCollection.java:249)
	at org.gradle.internal.classpath.generated.AbstractCompile_Adapter.access_set_setClasspath(AbstractCompile_Adapter.java:19)
	at com.android.build.gradle.tasks.JavaCompileUtils.configureProperties(JavaCompileUtils.kt:98)
	at com.android.build.gradle.tasks.JavaCompileCreationAction.configure(JavaCompile.kt:125)
	at com.android.build.gradle.tasks.JavaCompileCreationAction.configure(JavaCompile.kt:57)
	at com.android.build.gradle.internal.tasks.factory.TaskConfigurationActions.execute(TaskFactoryUtils.kt:94)
	at com.android.build.gradle.internal.tasks.factory.TaskConfigurationActions.execute(TaskFactoryUtils.kt:80)
```

It looks like the 4th element in [this AGP logic](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:build-system/gradle-core/src/main/java/com/android/build/gradle/tasks/JavaCompileUtils.kt;drc=f18bcd663ca8a9d6a6b9698d9091ad659861a0c6;l=98) can be null:
```
this.classpath = project.files(
    // classes(e.g. android.jar) that were previously passed through bootstrapClasspath need to be provided
    // through classpath
    creationConfig.global.bootClasspath,
    creationConfig.compileClasspath,
    creationConfig.getBuiltInKotlincOutput(),
    creationConfig.getBuiltInKaptArtifact(InternalArtifactType.BUILT_IN_KAPT_CLASSES_DIR), // <-- this one seems to be null
)
```

We should probably filter `null` in `UnresolvedItemsCollector` or find another solution or report to AGP.

--- a/platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy
+++ b/platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy
@@ -626,8 +626,7 @@ class FileCollectionIntegrationTest extends AbstractIntegrationSpec implements T
                     null
                 )
             }
-        """
-        )
+        """)
         when:
         run "foo"
 

--- a/platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy
+++ b/platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy
@@ -27,6 +27,7 @@ import spock.lang.Issue
 import static org.gradle.util.internal.TextUtil.escapeString
 
 class FileCollectionIntegrationTest extends AbstractIntegrationSpec implements TasksWithInputsAndOutputs {
+
     def "can use 'as' operator with #type"() {
         buildFile << """
             def fileCollection = files("input.txt")
@@ -604,6 +605,38 @@ class FileCollectionIntegrationTest extends AbstractIntegrationSpec implements T
         outputContains 'Transforming p2.zip'
         result.assertTasksExecuted ':p1:produce', ':p1:zip', ':p2:produce', ':p2:zip', ':merge'
         file('merge.txt').text == 'p1.zip,p2.zip'
+    }
+
+    def "ignores null as an argument for ConfigurableFileCollection.#api"() {
+        buildFile("""
+            abstract class FooTask extends DefaultTask {
+                @InputFiles
+                abstract ConfigurableFileCollection getIncoming()
+
+                @TaskAction
+                void foo() {
+                    println("Incoming: \${incoming.files.toSorted()}")
+                }
+            }
+
+            tasks.register("foo", FooTask) {
+                incoming.$api(
+                    provider { "a.txt" },
+                    provider { "b.txt" },
+                    null
+                )
+            }
+        """
+        )
+        when:
+        run "foo"
+
+        then:
+        def files = ["a.txt", "b.txt"]
+        outputContains("Incoming: ${files.collect { testDirectory.file(it) }.toSorted()}")
+
+        where:
+        api << ["setFrom", "convention", "from"]
     }
 
     private void withSubprojects(String... subprojects) {

--- a/platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java
+++ b/platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java
@@ -675,6 +675,9 @@ public class DefaultConfigurableFileCollection extends CompositeFileCollection i
             ImmutableList.Builder<Object> builder = ImmutableList.builderWithExpectedSize(items.size());
             boolean hasChanges = false;
             for (Object candidate : items) {
+                if (candidate == null) {
+                    continue;
+                }
                 if (candidate instanceof FileCollectionInternal) {
                     FileCollectionInternal newCollection = ((FileCollectionInternal) candidate).replace(original, supplier);
                     hasChanges |= newCollection != candidate;

--- a/subprojects/core-api/src/main/java/org/gradle/api/file/ConfigurableFileCollection.java
+++ b/subprojects/core-api/src/main/java/org/gradle/api/file/ConfigurableFileCollection.java
@@ -42,14 +42,14 @@ public interface ConfigurableFileCollection extends FileCollection, HasConfigura
     /**
      * Sets the source paths for this collection. The given paths are evaluated as per {@link org.gradle.api.Project#files(Object...)}.
      *
-     * @param paths The paths.
+     * @param paths The paths. {@code null} values are ignored.
      */
     void setFrom(Iterable<?> paths);
 
     /**
      * Sets the source paths for this collection. The given paths are evaluated as per {@link org.gradle.api.Project#files(Object...)}.
      *
-     * @param paths The paths.
+     * @param paths The paths. {@code null} values are ignored.
      */
     void setFrom(Object... paths);
 
@@ -60,7 +60,7 @@ public interface ConfigurableFileCollection extends FileCollection, HasConfigura
      * If, at the time this method is invoked, the set of source paths for this collection is empty, the convention will be used
      * to resolve this file collection.
      *
-     * @param paths The paths.
+     * @param paths The paths. {@code null} values are ignored.
      * @return this collection
      *
      * @since 8.8
@@ -75,7 +75,7 @@ public interface ConfigurableFileCollection extends FileCollection, HasConfigura
      * If, at the time this method is invoked, the set of source paths for this collection is empty, the convention will be used
      * to resolve this file collection.
      *
-     * @param paths The paths.
+     * @param paths The paths. {@code null} values are ignored.
      * @return this collection
      *
      * @since 8.8
@@ -86,7 +86,7 @@ public interface ConfigurableFileCollection extends FileCollection, HasConfigura
     /**
      * Adds a set of source paths to this collection. The given paths are evaluated as per {@link org.gradle.api.Project#files(Object...)}.
      *
-     * @param paths The files to add.
+     * @param paths The files to add. {@code null} values are ignored.
      * @return this
      */
     ConfigurableFileCollection from(Object... paths);

